# PhyG-Pi 
**PhyG** is a flexible modular **Phy**lo**G**enomic-**Pi**peline (PhyG-Pi), with a suite of tools and modules to aid in various phylogenomic related tasks. Development of PhyG was motivated by the rapid generation of genome-scale datasets, leading to a need for efficient and reproducible phylogenomic approaches that can provide a 'streamlined' pipeline for quick analyses, while offering control over each major step in developing phylogenomic matrices. PhyG also includes **alignment-free**, **natural language processing**, and **machine learning** approaches for reproducible alignment refinement and putative contamination identification from particularly troublesome datasets (e.g., predator-prey transcriptomes).
 
PhyG is designed to provide support for various dataset cleaning (e.g., decontamination) and phylogenetic reconstruction tasks, including generating species trees and single-gene trees for various evolutionary analyses (e.g., evolutionary rates and searching for lateral gene transfers).

 Key features of PhyG
-----------------------
* __Custom Gene Family Databases__: Support for adding taxa to pre-existing gene family databases. PhyG **will** provide several preformatted common phylogenomic databases (from eukaryotic tree of life studies).
* __Efficient Alignment Refinement__: Multiple avenues for multi-sequence alignment, including use of reproducible natural language processing and machine-learning based filtering approaches.
* __Quick Contamination Evaluation__: Several approaches for inferring putative contamination (eukaryote-eukaryote) to aid in further cleaning of tricky datasets.
* __Streamlining Phylogenomics__: Provides a streamlined 'start-to-finish' means to generate single-gene phylogenies and/or gene tree-species tree phylogenomic reconstructions.

See the [PhyG wiki](https://github.com/xxmalcala/PhyG/wiki) for more details about current functionality and installation.

**Currently targeting early December 2024 for full public release.**
